content-src/components/SiteIcon/SiteIcon.js

```diff
@@ -0,0 +1,95 @@
+const React = require("react");
+const classNames = require("classnames");
+
+const DEFAULT_FALLBACK_BG_COLOR = [200, 200, 200];
+const MIN_RESOLUTION = 2;
+const FAVICON_SIZE = 16;
+
+function getSiteImage(site, width) {
+  const {favicon_url, icons} = site;
+  if (icons && icons.length) {
+    return icons[0].url;
+  } else if (favicon_url && width / FAVICON_SIZE <= MIN_RESOLUTION) {
+    return favicon_url;
+  } else {
+    return false;
+  }
+}
+
+const SiteIconImage = React.createClass({
+  render() {
+    // TODO: Do something more sophisticated than choose the first icon in the array
+    // const imageUrl = this.props.icons[0].url;
+    return (<div className="site-icon-image" ref="image"
+      style={{backgroundImage: `url(${getSiteImage(this.props, this.props.width)})`}}>
+    </div>);
+  }
+});
+
+SiteIconImage.propTypes = {
+  icons: React.PropTypes.arrayOf(React.PropTypes.shape({
+    url: React.PropTypes.string.isRequired
+  })),
+  favicon_url: React.PropTypes.string,
+  width: React.PropTypes.number
+};
+
+const SiteIconFallback = React.createClass({
+  render() {
+    const {favicon_colors, title, provider_name, provider_display} = this.props;
+    const letter = (provider_name || provider_display || title)[0];
+    // TODO: do something more sophisticated than choose the first color
+    const color = favicon_colors && favicon_colors[0] && favicon_colors[0].color || DEFAULT_FALLBACK_BG_COLOR;
+    return (<div className="site-icon-fallback" ref="fallback"
+      style={{backgroundColor: `rgb(${color.join(", ")})`}}>
+      {letter}
+    </div>);
+  }
+});
+
+SiteIconFallback.propTypes = {
+  favicon_colors: React.PropTypes.arrayOf(React.PropTypes.shape({
+    color: React.PropTypes.arrayOf(React.PropTypes.number)
+  })),
+  provider_name: React.PropTypes.string,
+  provider_display: React.PropTypes.string,
+  title: React.PropTypes.string
+};
+
+const SiteIcon = React.createClass({
+  getDefaultProps() {
+    return {
+      height: 100,
+      width: 100
+    };
+  },
+  render() {
+    const {site, width, height} = this.props;
+    const Icon = getSiteImage(site, width) ? SiteIconImage : SiteIconFallback;
```

Comment:
I don't know if we should ignore this now, but getSiteImage() returns potentially different types that get coerced here - which is confusing. It would be better if `getSiteImage()` returned only a boolean.